src/services/redisUserMetadataStorage.ts

```diff
@@ -1,60 +1,122 @@
 import * as redis from "redis";
 
-import { Either, isLeft, isRight, left, right } from "fp-ts/lib/Either";
+import { Either, isLeft, left, right, toError } from "fp-ts/lib/Either";
 import { ReadableReporter } from "italia-ts-commons/lib/reporters";
 import { UserMetadata } from "../../generated/backend/UserMetadata";
 import { User } from "../types/user";
 import { log } from "../utils/logger";
 import { IUserMetadataStorage } from "./IUserMetadataStorage";
 import RedisStorageUtils from "./redisStorageUtils";
 
+import { Sema } from "async-sema";
+import {
+  fromEither,
+  fromPredicate,
+  taskify,
+  tryCatch
+} from "fp-ts/lib/TaskEither";
+
 const userMetadataPrefix = "USERMETA-";
 export const metadataNotFoundError = new Error("User Metadata not found");
 export const invalidVersionNumberError = new Error("Invalid version number");
+export const concurrentWriteRejectionError = new Error(
+  "Concurrent write operation"
+);
 
 /**
  * Service that manages user metadata stored into Redis database.
  */
 export default class RedisUserMetadataStorage extends RedisStorageUtils
   implements IUserMetadataStorage {
+  private activeClients: Set<string> = new Set();
+  private mutex: Sema = new Sema(1);
   constructor(private readonly redisClient: redis.RedisClient) {
     super();
   }
 
   /**
    * {@inheritDoc}
-   *
-   * This method doesn't support atomic operations on concurrency scenario.
-   * Story https://www.pivotaltracker.com/story/show/167064659
+   * This method uses Optimistic Lock to prevent race condition
+   * during write operations of user metadata
+   * @see https://github.com/NodeRedis/node_redis#optimistic-locks
    */
   public async set(
     user: User,
     payload: UserMetadata
   ): Promise<Either<Error, boolean>> {
-    const getUserMetadataResult = await this.loadUserMetadataByFiscalCode(
-      user.fiscal_code
-    );
-    if (
-      isRight(getUserMetadataResult) &&
-      getUserMetadataResult.value.version !== payload.version - 1
-    ) {
-      return left(invalidVersionNumberError);
-    }
-    if (
-      isLeft(getUserMetadataResult) &&
-      getUserMetadataResult.value !== metadataNotFoundError
-    ) {
-      return left(getUserMetadataResult.value);
+    // In order to work properly, optimistic lock needs to be initialized on different
+    // redis client instances @see https://github.com/NodeRedis/node_redis/issues/1320#issuecomment-373200541
+    await this.mutex.acquire();
```

Comment:
I suggest to extract this logic into a kind of client pool class

Reply:
a couple of suggestions on `async-sema`:

1. review [the defaults](https://www.npmjs.com/package/async-sema#constructornr--initfn-pausefn-resumefn-capacity-), are they good for us?
1. use `tryAcquire` in place of `acquire` to avoid accumulating "concurrent threads", if the mutex can't be acquired, fail with an HTTP 429

Reply:
> I suggest to extract this logic into a kind of client pool class

can you please elaborate a little bit more ? do you mean to create (at startup) a pool of N redisclients to reuse and change the mutex(1) to mutex(N) ?

do you think we can merge this PR as-is and implement the pooling in another PR ? (after the beta release date)